### ISSUE TITLE
Rendering Automation

### ISSUE THREAD

**Issue by harrisse** [ephtracy/ephtracy.github.io#67]
I'm not sure if this is already possible, but I haven't been able to find references to it (besides http://drinkdecaf.com/magicavoxel_animate). My use case is that I have a bunch of .vox models and would like to produce nice rendered images of each of them from a variety of camera angles. My problem is that rendering each image is slow, MagicaVoxel has to stay in the foreground, and I have to wait for rendering to finish from each angle every time I change my model. My proposal is that I'd like to be able to call MagicaVoxel.exe from the command line, passing in a camera transform and projection, image size, pixel samples, etc and it would render the image as if I had entered those settings in the UI. Bonus points if I can pick a transparent background 😄

**Comment by integerpoet** [comment]
I came here to say roughly the same thing. The spirit of this idea seems good and I expect it would make communicating with management and marketing people easier.

Of course, the app already has its own command line (which I have not explored much), so it would make more sense to put it there than on an EXE command line. And I would prefer a GUI way to do this.

**Comment by harrisse** [comment]
Disregard the transparent background section, just learned about the alpha channel button!

@integerpoet I think the problem for me with the built in command line is that it doesn't offer scripting abilities, and I suspect that scripting is probably out of scope for what MagicaVoxel is supposed to do and would be a lot more work than exposing its internal commands to the command line and using an external scripting language.

Ultimately ephtracy is the only one who knows what'd work best with MagicaVoxel's current state / design goals though (if this kind of functionality is even in scope for MV).

**Comment by integerpoet** [comment]
I’m not sure whether it’s more or less difficult to enhance scriptability driven by any particular command line, but the app is already somewhat scriptable via the internal command line. See some of the third party shaders listed on the web site.

**Comment by harrisse** [comment]
Right, shaders can create programmatic models, but can't control stuff like camera rotation, sun rotation / intensity, render resolution, etc (at least AFAIK, that would be pretty crazy if so). For example, what if I wanted to render the same model from each of the 8 45 degree angles, each at 3 different sun angles. That kind of scripting needs constructs like for loops, variables, etc that all take quite a bit more work than commands. It'd also be repeating previously done work without much benefit. One option might be introducing something like LUA for internal scripting, but I don't think this gets much benefit over exposing the internal command line and using an external scripting language.

**Comment by harrisse** [comment]
Command lines are also not the most intuitive scripting environment. The reason I suggest exposing this through an external command line is that external command lines are invokable through external scripting languages. For large automation workloads a real scripting language really does have a ton of benefits over a command prompt.

**Comment by pims** [comment]
Being able to specify rendering options on the command line would make MV a million times more useful.

Does anyone know of a scriptable voxel renderer as a temporary alternative until MV is capable of this?

**Comment by bursaar** [comment]
I'd like to throw my hat in here too – it'd be incredible to have the ability to automate rendering, it'd open up so many possibilities!

**Comment by thelotus819** [comment]
I don't know if I can talk about other projects where, I'm doing an isometric game and needed a lot of images. I ended up using https://github.com/tommyettinger/IsoVoxel the end result isn't as good as MV but it automate the generation of images for development.

**Comment by RobQuistNL** [comment]
This would be an awesome feature.

**Comment by SamHasler** [comment]
I looked at editing the config files to add buttons to the interface to automate following a path, but there didn't appear to be a way to automate moving the camera in the way I wanted as the only commands available were the pre-recorded positions.

One way to do it with the current limitations would be to write a program/script that would:

1. Insert a camera positions into the camera save file
2. Send keystokes to MV to move the camera to that saved position
3. Send keystokes to MV to render a frame
4. Somehow detect that MV has finished rendering.
5. Start again at 1 but inserting a new position in a different camera slot.

**Comment by juancampa** [comment]
Yeah this would be awesome! @ephtracy any chance we could get some way of automating rendering from the command line?

**Comment by drbh** [comment]
+1 this would be awesome

**Comment by cobycloud** [comment]
> I looked at editing the config files to add buttons to the interface to automate following a path, but there didn't appear to be a way to automate moving the camera in the way I wanted as the only commands available were the pre-recorded positions.
> 
> One way to do it with the current limitations would be to write a program/script that would:
> 
> 1. Insert a camera positions into the camera save file
> 2. Send keystokes to MV to move the camera to that saved position
> 3. Send keystokes to MV to render a frame
> 4. Somehow detect that MV has finished rendering.
> 5. Start again at 1 but inserting a new position in a different camera slot.

yes, it is possible to leverage pyautogui for this effect, but in general it's bad practice to rely on automating peripherals as such. Every good software should be designed for both human and machine users. We all love the simplicity as human users, but the true strength of the software will stem from its ability to integrate with machine users - be it a python script or react app
